components/sup/src/event/stream.rs

```diff
@@ -3,31 +3,81 @@ use crate::event::{Error,
                    EventStreamConnectionInfo,
                    Result};
 use futures::sync::mpsc as futures_mpsc;
+use nats::Client;
 use std::{sync::mpsc as std_mpsc,
-          thread};
+          thread,
+          time::Duration};
+use tokio::{prelude::Stream,
+            runtime::current_thread::Runtime};
 
 /// All messages are published under this subject.
 const HABITAT_SUBJECT: &str = "habitat";
 
 pub(super) fn init_stream(conn_info: EventStreamConnectionInfo) -> Result<EventStream> {
+    // TODO (DM): This cannot be unbounded. We need backpressure. If the connection is down when we
+    // try to publish we try to reconnect this can be time consuming so we can easily get
+    // behind.
     let (event_tx, event_rx) = futures_mpsc::unbounded();
     let (sync_tx, sync_rx) = std_mpsc::sync_channel(0); // rendezvous channel
 
     let EventStreamConnectionInfo { name,
                                     verbose,
                                     cluster_uri,
-                                    cluster_id,
-                                    auth_token,
+                                    // TODO (DM): The nats client we are using does not support
+                                    // auth tokens and will need to be patched
+                                    auth_token: _,
                                     connect_method, } = conn_info;
-    let connect_method_is_timeout = connect_method.is_timeout();
+    let connection_is_timeout = connect_method.is_timeout();
+
+    // Note: With the way we are using the client, we will not respond to pings from the server
+    // (https://nats-io.github.io/docs/nats_protocol/nats-protocol.html#pingpong).
+    // Instead, we rely on publishing events to keep us connected to the server. This is completely
+    // valid according to the protocol, and in fact, the server will not send pings if there is
+    // other traffic from a client. If we have no events for an extended period of time, we will
+    // be automatically disconnected (because we do not respond to pings). When the next event comes
+    // in we will try to reconnect.
```

Comment:
If we later decide we need to participate in the `PING/PONG` interaction, we can always compose the `event_handler` future with something like an [`Interval`](https://docs.rs/tokio-timer/0.2.11/tokio_timer/struct.Interval.html) to send a `PING` periodically if we don't get an actual event to send.

Reply:
But just as a point of clarification: if we disconnect, the underlying NATS client library will handle the reconnect for us, correct? That's nothing that we need to manage ourselves?

Reply:
> But just as a point of clarification: if we disconnect, the underlying NATS client library will handle the reconnect for us, correct?

Correct, if a publish fails the library will try and reconnect and then republish.